src/OPTokenClaim.sol

```diff
@@ -54,62 +72,121 @@ contract OPTokenClaim is Ownable {
         emit ClaimExtended(months);
     }
 
-    function claimOP(address account) external {
+    /* ========== VIEWS ========== */
+
+    /// returns current epoch number
+    function currentEpoch() external view returns (uint256 epochNumber) {
+        unchecked {
+            epochNumber = (block.timestamp - config.start) / 30 days;
+        }
+    }
+
+    /// returns last epoch where claims are open
+    function maxEpoch() external view returns (uint256) {
+        return config.maxEpoch;
+    }
+
+    /// calculates reward for account at given epoch
+    function rewardAtEpoch(address account, uint256 epoch) external view returns (uint256) {
+        return _calcReward(account, epoch);
+    }
+
+    /// returns number of subscribed accounts at epoch
+    function accountsAtEpoch(uint256 epoch) external view returns (uint128) {
+        return epochs[epoch].numAccounts;
+    }
+
+    /// returns total subscribed EXP at epoch
+    function totalEXPAtEpoch(uint256 epoch) external view returns (uint128) {
+        return epochs[epoch].totalEXP;
+    }
+
+    /// returns subscribed EXP of account at given epoch and if account claimed
+    function accountInfoAtEpoch(address account, uint256 epoch) external view returns (uint256, bool) {
+        return (epochs[epoch].addressToEXP[account], epochs[epoch].addressClaimed[account]);
+    }
+
+    /* ========== USER FUNCTIONS ========== */
+
+    /// subscribe to reward distribution for current epoch
+    function subscribe(address account) external {
         Config memory _config = config;
 
         // epoch 0 is the first
-        uint256 epoch = currentEpoch(_config);
+        uint256 epoch = _currentEpoch(_config);
+        require(epoch < _config.maxEpoch, "claims ended");
+        require(epochs[epoch].addressToEXP[account] == 0, "already subscribed for this epoch");
+
+        uint256 expBalance = EXP.balanceOf(account);
+        require(expBalance > 0, "address has no exp");
+
+        // cap at MAX_EXP = 99 EXP
+        if (expBalance > MAX_EXP) {
+            expBalance = MAX_EXP;
+        }
+
+        epochs[epoch].addressToEXP[account] = expBalance;
+        epochs[epoch].totalEXP += uint128(expBalance);
+        epochs[epoch].numAccounts++;
+
+        emit Subscribed(account, epoch, expBalance);
```

Comment:
```suggestion
        uint256 epochNum = _currentEpoch(_config);
        require(epochNum < _config.maxEpoch, "claims ended");
        
        EpochInfo memory epoch = epochs[epochNum];
        
        require(epoch.addressToEXP[account] == 0, "already subscribed for this epoch");

        uint256 expBalance = EXP.balanceOf(account);
        require(expBalance > 0, "address has no exp");

        // cap at MAX_EXP = 99 EXP
        if (expBalance > MAX_EXP) {
            expBalance = MAX_EXP;
        }
 
        unchecked {
            epoch.addressToEXP[account] = expBalance;
            epoch.totalEXP += uint128(expBalance);
            epoch.numAccounts++;
        }

        epochs[epochNum] = epoch;

        emit Subscribed(account, epoch, expBalance);
```

Reply:
ok actually testing this out locally, it doesn't work because of the nested mapping. Let me think some more about it

Reply:
opened a shared branch here with your commits and some of my changes here: https://github.com/ethernautdao/ethernaut-op-claim/pull/3